packages/std/src/math/int512.rs

```diff
@@ -0,0 +1,1350 @@
+            words[5].to_be_bytes(),
+            words[4].to_be_bytes(),
+            words[3].to_be_bytes(),
+            words[2].to_be_bytes(),
+            words[1].to_be_bytes(),
+            words[0].to_be_bytes(),
+        ];
+        unsafe { std::mem::transmute::<[[u8; 8]; 8], [u8; 64]>(words) }
+    }
+
+    /// Returns a copy of the number as little endian bytes.
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    pub const fn to_le_bytes(self) -> [u8; 64] {
+        let bits = self.0.to_bits();
+        let words = bits.digits();
+        let words = [
+            words[0].to_le_bytes(),
+            words[1].to_le_bytes(),
+            words[2].to_le_bytes(),
+            words[3].to_le_bytes(),
+            words[4].to_le_bytes(),
+            words[5].to_le_bytes(),
+            words[6].to_le_bytes(),
+            words[7].to_le_bytes(),
+        ];
+        unsafe { std::mem::transmute::<[[u8; 8]; 8], [u8; 64]>(words) }
+    }
+
+    #[must_use]
+    pub const fn is_zero(&self) -> bool {
+        self.0.is_zero()
+    }
+
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    pub fn pow(self, exp: u32) -> Self {
+        Self(self.0.pow(exp))
+    }
+
+    pub fn checked_add(self, other: Self) -> Result<Self, OverflowError> {
+        self.0
+            .checked_add(other.0)
+            .map(Self)
+            .ok_or_else(|| OverflowError::new(OverflowOperation::Add, self, other))
+    }
+
+    pub fn checked_sub(self, other: Self) -> Result<Self, OverflowError> {
+        self.0
+            .checked_sub(other.0)
+            .map(Self)
+            .ok_or_else(|| OverflowError::new(OverflowOperation::Sub, self, other))
+    }
+
+    pub fn checked_mul(self, other: Self) -> Result<Self, OverflowError> {
+        self.0
+            .checked_mul(other.0)
+            .map(Self)
+            .ok_or_else(|| OverflowError::new(OverflowOperation::Mul, self, other))
+    }
+
+    pub fn checked_pow(self, exp: u32) -> Result<Self, OverflowError> {
+        self.0
+            .checked_pow(exp)
+            .map(Self)
+            .ok_or_else(|| OverflowError::new(OverflowOperation::Pow, self, exp))
+    }
+
+    pub fn checked_div(self, other: Self) -> Result<Self, DivisionError> {
+        if other.is_zero() {
+            return Err(DivideByZeroError::new(self).into());
+        }
+        self.0
+            .checked_div(other.0)
+            .map(Self)
+            .ok_or_else(|| OverflowError::new(OverflowOperation::Div, self, other).into())
+    }
+
+    pub fn checked_div_euclid(self, other: Self) -> Result<Self, DivisionError> {
+        if other.is_zero() {
+            return Err(DivideByZeroError::new(self).into());
+        }
+        self.0
+            .checked_div_euclid(other.0)
+            .map(Self)
+            .ok_or_else(|| OverflowError::new(OverflowOperation::Div, self, other).into())
+    }
```

Comment:
So instead of wrapping the existing `OverflowError` in the DivisionError, we use a variant with just the two operands?
We can do that for the `DivisionError`, but for `checked_neg` we would still have to add `OverflowOperation::Neg` (unless we duplicate the whole error just for that)

Reply:
Yeah. My thinking is this: `OverflowError` is not great designed anyways, especially with the string allocations. We don't need to introduce more usages of it. Now what we do right now is to put a very specific error case (division overflows) into a more open container where all the other  `OverflowOperation` cases are impossible to have. So calling code would have to assert `OverflowOperation::Div`.

I'd just go for

```rust
#[derive(Error, Debug, PartialEq, Eq)]
pub enum DivisionError {
    #[error("Divide by zero")]
    DivideByZero,

    #[error("Overflow in division")]
    Overflow,
}
```